doc/warm-reboot/view_switch.md

```diff
@@ -0,0 +1,179 @@
+    # Current behavior
+    SADD KeySet key
+    SADD DelKeySet key
+    DEL StateHash:key
+    PUBLISH Channel
+}
+```
+
+#### `apply_temp_view()`
+
+To apply temporary view, there are several steps:
+1. Drop all pending operations by clearing KeySet, DelKeySet, and StateHash
+2. Dump content of current view
+3. Compare current view and target view, generate a minimal set of `set()` and `del()` operations
+4. Write to KeySet, DelKeySet, and StateHash accordingly, publish the change to the channel, and unset `m_tempViewActive` flag.
+
+```
+# Drop all pending operations first
+DEL KeySet
+DEL DelKeySet
+for key in KEYS StateHash:*
+    DEL StateHash:key
+end
+
+TableDump tableDump;
+dump(tableDump);
+
+std::vector<std::string> keysToSet;
+std::vector<std::string> keysToDel;
+
+// For all old objects
+for (auto const& [key, fieldValueMap] : tableDump)
+{
+    // DEL is needed if object does not exist in new state, or any field is not presented in new state
+    // SET is almost always needed, unless old state and new state exactly match each other
+    //     (All old fields exists in new state, values match, and there is no additional field in new state)
+    if (m_tempViewState.find(key) == m_tempViewState.end())         // Key does not exist in new view
+    {
+        keysToDel.emplace_back(key);
+        continue;
+    }
+    newFieldValueMap = m_tempViewState[key];
+    bool needDel = false;
+    bool needSet = false;
+    for (auto const& [field, value] : fieldValueMap)
+    {
+        if (newFieldValueMap.find(field) == newFieldValueMap.end()) // Field does not exist in new view
+        {
+            needDel = true;
+            needSet = true;
+            break;
+        }
+        if (newFieldValueMap[field] != value)                       // Field value changed
+        {
+            needSet = true;
+        }
+    }
+    if (newFieldValueMap.size() > fieldValueMap.size())             // New field added
+    {
+        needSet = true;
+    }
+    if (needDel) keysToDel.emplace_back(key);
+    if (needSet) keysToSet.emplace_back(key);
+    else newFieldValueMap.erase(key);                               // If exactly match, no need to sync new state to StateHash in DB
+}
+
+// Generate a flat arg list from m_tempViewState to be written to m_tempViewActive
+std::vector<std::string> keyFieldValueStates;
+prepareStateHashArgument(keyFieldValueStates, m_tempViewState);
+
+m_tempViewActive = false;
+m_tempViewState.clear() 
+
+for key, field, value in keyFieldValueStates do
+    HSET StateHash:key field value
+end
+for key in keysToSet do
+    SADD KeySet key
+end
+for key in keysToDel do
+    SADD DelKeySet key
+end
+PUBLISH Channel
+```
+
+#### `pop()` in ConsumerStateTable
+As `apply_temp_view()` is translating view switching operation into a set of standard `set()` and `del()` operations,
+no change is needed for `pop()` and any other functions in ConsumerStateTable.
+
+
+### Q&A
+
+#### How about storing temporary view in DB instead?
```

Comment:
locking mechanism? should return fail when the second producer try to do view switching when there is already one going on.